diskwriter.go

```diff
@@ -345,3 +398,83 @@ func nextSuffix() string {
 	randmu.Unlock()
 	return strconv.Itoa(int(1e9 + r%1e9))[1:]
 }
+
+// files returns the list of changes that change
+// regular files (e.g. where the file reference is non-nil)
+func (r fileChanges) files() (res []fileChange) {
+	for _, change := range r {
+		if change.file != nil {
+			res = append(res, change)
+		}
+	}
+	return res
+}
+
+type fileChanges []fileChange
+
+type fileChange struct {
+	// newPath specifies the new absolute path to the file/directory
+	// in case of a rename
+	newPath string
+	// destPath is the absolute path to the file
+	// in destination directory
+	destPath string
+	rename   bool
+	// typChange indicates that a file is being switched to a directory
+	// or vice versa and the old location should be cleaned up
+	typChange bool
+	// file optionally specifies the file
+	file *os.File
+	stat types.Stat
+}
+
+func (r *multiWriter) Close() (err error) {
+	var errs []error
+	for _, wc := range r.wcs {
+		if err := wc.Close(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return newMultiErr(errs...)
+}
+
+func newMultiWriter(changes []fileChange) *multiWriter {
+	var ws []io.Writer
+	var wcs []io.WriteCloser
+	for _, change := range changes {
+		ws = append(ws, change.file)
+		wcs = append(wcs, change.file)
+	}
+	return &multiWriter{
+		Writer: io.MultiWriter(ws...),
+		wcs:    wcs,
+	}
+}
+
+type multiWriter struct {
+	io.Writer
+	wcs []io.WriteCloser
+}
+
+func newMultiErr(errs ...error) error {
```

Comment:
I did not attempt to introduce another dependency on something like [go-multierror](github.com/hashicorp/go-multierror) so this is ad-hoc multiple errors wrapper.